File: internal/scaffold/orchestration.go
package scaffold

import (
	"io/fs"
	"log"
	"path/filepath"

	"github.com/ricardoborges-teachable/ai-setup/internal/conflict"
	"github.com/ricardoborges-teachable/ai-setup/internal/files"
	"github.com/ricardoborges-teachable/ai-setup/internal/types"
)

// ScaffoldOrchestration installs orchestration definitions (chains, teams,
// workflows, skills) to .ai/orchestration/.
// Ported from src/scaffold/orchestration.ts.
func ScaffoldOrchestration(targetDir string, libFS fs.FS, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	if !files.ExistsFS(libFS, "orchestration") {
		return nil
	}

	targetRoot := filepath.Join(targetDir, ".ai", "orchestration")
	if err := files.EnsureDir(targetRoot); err != nil {
		return err
	}

	return copyOrchestrationTreeFS(libFS, "orchestration", targetRoot, targetDir, fileRecords, strategy, perFileOverrides)
}

// copyOrchestrationTreeFS recursively copies files from the library FS to the target directory.
func copyOrchestrationTreeFS(libFS fs.FS, sourceRelDir, currentTargetDir, targetDir string, fileRecords *[]types.TrackedFile, strategy types.ConflictStrategy, perFileOverrides map[string]types.ConflictStrategy) error {
	entries, err := files.ReadDirFS(libFS, sourceRelDir)
	if err != nil {
		return nil // directory doesn't exist in FS, skip
	}

	for _, entry := range entries {
		sourceRelPath := sourceRelDir + "/" + entry.Name()
		targetPath := filepath.Join(currentTargetDir, entry.Name())

		if entry.IsDir() {
			if err := files.EnsureDir(targetPath); err != nil {
				return err
			}
			if err := copyOrchestrationTreeFS(libFS, sourceRelPath, targetPath, targetDir, fileRecords, strategy, perFileOverrides); err != nil {
				return err
			}
			continue
		}

		relPath, err := filepath.Rel(targetDir, targetPath)
		if err != nil {
			relPath = targetPath
		}

		action, err := conflict.ApplyStrategy(targetPath, strategy, perFileOverrides, targetDir)
		if err != nil {
			return err
		}
		if action == "skip" {
			log.Printf("Skipping existing file: %s", relPath)
			continue
		}

		// Read from library FS and write to target.
		data, err := files.ReadFS(libFS, sourceRelPath)
		if err != nil {
			log.Printf("Warning: could not read %s: %v", sourceRelPath, err)
			continue
		}

		if err := files.WriteFile(targetPath, data, 0o644); err != nil {
			return err
		}

		hash, _ := files.FileHash(targetPath)
		*fileRecords = append(*fileRecords, types.TrackedFile{
			Path:   relPath,
			Hash:   hash,
			Source: sourceRelPath,
			Owner:  types.FileOwnerLibrary,
		})
	}

	return nil
}

File: packages/ai-setup-ts/src/scaffold/orchestration.ts
import path from 'node:path'
import { discoverExtensions } from '../extensions/discovery.js'
import type { FeatureFlags } from '../store/schema.js'
import type { ConflictStrategy, FileRecord } from '../types.js'
import { applyStrategy } from '../utils/conflict-strategy.js'
import { copyFile, ensureDir, fileExists, fileHash, isDirectory, listDir } from '../utils/files.js'

export interface ScaffoldOrchestrationOptions {
  targetDir: string
  libraryDir: string
  fileRecords: FileRecord[]
  strategy: ConflictStrategy
  perFileOverrides: Map<string, ConflictStrategy>
  features?: Partial<FeatureFlags>
}

export async function scaffoldOrchestration(opts: ScaffoldOrchestrationOptions): Promise<void> {
  const sourceRoot = path.join(opts.libraryDir, 'orchestration')
  if (!fileExists(sourceRoot)) return

  const targetRoot = path.join(opts.targetDir, '.ai', 'orchestration')
  ensureDir(targetRoot)

  await copyTree({
    sourceRoot,
    currentSourceDir: sourceRoot,
    currentTargetDir: targetRoot,
    targetDir: opts.targetDir,
    recordSourceRoot: opts.libraryDir,
    fileRecords: opts.fileRecords,
    strategy: opts.strategy,
    perFileOverrides: opts.perFileOverrides,
    features: opts.features,
  })

  for (const extension of discoverExtensions(opts.targetDir)) {
    const extensionOrchestrationDir = path.join(extension.path, 'orchestration')
    if (!isDirectory(extensionOrchestrationDir)) continue

    for (const category of ['chains', 'teams', 'workflows'] as const) {
      const sourceDir = path.join(extensionOrchestrationDir, category)
      if (!isDirectory(sourceDir)) continue

      const targetDir = path.join(targetRoot, category)
      ensureDir(targetDir)

      await copyTree({
        sourceRoot: sourceDir,
        currentSourceDir: sourceDir,
        currentTargetDir: targetDir,
        targetDir: opts.targetDir,
        recordSourceRoot: extension.path,
        fileRecords: opts.fileRecords,
        strategy: opts.strategy,
        perFileOverrides: opts.perFileOverrides,
        features: undefined,
      })
    }
  }
}

interface CopyTreeOptions {
  sourceRoot: string
  currentSourceDir: string
  currentTargetDir: string
  targetDir: string
  recordSourceRoot: string
  fileRecords: FileRecord[]
  strategy: ConflictStrategy
  perFileOverrides: Map<string, ConflictStrategy>
  features: Partial<FeatureFlags> | undefined
}

async function copyTree(opts: CopyTreeOptions): Promise<void> {
  const entries = listDir(opts.currentSourceDir)

  for (const entry of entries) {
    const sourcePath = path.join(opts.currentSourceDir, entry)
    const sourceRecordPath = path.relative(opts.recordSourceRoot, sourcePath)
    if (sourceRecordPath === path.join('orchestration', 'chains', 'feature-adversarial.json')) {
      continue
    }
    const targetPath = path.join(opts.currentTargetDir, entry)

    if (isDirectory(sourcePath)) {
      ensureDir(targetPath)
      await copyTree({
        ...opts,
        currentSourceDir: sourcePath,
        currentTargetDir: targetPath,
      })
      continue
    }

    const relPath = path.relative(opts.targetDir, targetPath)
    const action = applyStrategy(targetPath, opts.strategy, opts.perFileOverrides, opts.targetDir)

    if (action === 'skip') {
      console.warn(`⚠️  Skipping existing file: ${relPath}`)
      continue
    }

    const selectedSourcePath = selectOrchestrationSource(sourcePath, opts)
    copyFile(selectedSourcePath, targetPath)
    opts.fileRecords.push({
      path: relPath,
      hash: fileHash(targetPath),
      source: path.relative(opts.recordSourceRoot, selectedSourcePath),
      owner: 'library',
    })
  }
}

function selectOrchestrationSource(sourcePath: string, opts: CopyTreeOptions): string {
  const sourceRecordPath = path.relative(opts.recordSourceRoot, sourcePath)
  if (sourceRecordPath !== path.join('orchestration', 'chains', 'feature.json') || !opts.features?.adversarialDesign) {
    return sourcePath
  }

  const adversarialSourcePath = path.join(path.dirname(sourcePath), 'feature-adversarial.json')
  return fileExists(adversarialSourcePath) ? adversarialSourcePath : sourcePath
}
